### DISCUSSION TITLE
How to add dof(s) for the basis of 'Shape' object

### DISCUSSION THREAD

**Discussion by Beta-y** [BJCaasenbrood/SorotokiCode#21] [General]
Hi, Dr. Caasenbrood. I have been working hard on study this kit for several days via investigating the codes and your corresponding papers and thesis. It is really an excellent tool for simulating soft stuffs. 

Actually, I am wondering **if there is a way to add degree of freedom(s) for the basis of `Shape` object so that it can be directly drived according to some controller or being under-actuated while no control input provided**, of course. I did notice that there is a function named `setBase` which can change the basis location by setting an exact pose. But it seems like it can not be drived by using a hdle function like `addControl`. The following is an example using PCC model with only one segment, in which I increase the $x$-angle of basis according to time $t$ using `setBase` function:
![soft](https://github.com/BJCaasenbrood/SorotokiCode/assets/58842921/112101a8-e5ed-4677-8258-2c0891f6637b)

Could you please give me some advices? Thank your pretty much! 😃

**Comment by BJCaasenbrood** [comment]
I've had the idea of free floating Cosserat beams for a while but never gotten the time to implement it. I do think most of the ingredients are there, except some minor computations here and there. I recently stumbled upon the paper from [Caradonna et al. (2024)](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=10502208) (see Section 3B), maybe this can help?

The paper can be found here [Model_and_Control_of_R-Soft_Inverted_Pendulum.pdf](https://github.com/BJCaasenbrood/SorotokiCode/files/15138700/Model_and_Control_of_R-Soft_Inverted_Pendulum.pdf)

I'll have to look into it to make a working implementation...

**Reply by BJCaasenbrood** [reply]
@Beta-y I am thinking about posting an issue/future enhancement.

**Reply by Beta-y** [reply]
Ohhh! Thanks Dr.! That is exactly what I am looking for! It is really fresh out of the oven. I will take a deep look into this paper, as well as looking forward to the on-coming enhancement! ❤

**Reply by BJCaasenbrood** [reply]
@Beta-y I currently have a working implementation. I'll be submitting the code tomorrow. You can construct a free-floating Cosserat beam using `Shapes(pod, dof, dof_base)`. An example is for instance:

```matlab
M   = 8 % number of modes
pod = chebyspace(50,M);
dof_beam = [0,M,0,0,0,0];
dof_base = [0,1,0,0,0,0]; % only leave rotation free

shp = Shapes(pod, dof_beam, dof_base)

```

Here is the result with the system under gravity.

![soft_pendulum](https://github.com/BJCaasenbrood/SorotokiCode/assets/36262141/a6c08496-3b7f-4271-a49b-d1a3539f6ffa)

**Reply by BJCaasenbrood** [reply]
The plotting and simulation is done with this code:

```matlab
while shp.solver.Time < shp.solver.TimeHorizon
    shp = shp.update();
    plotcurve(shp);
end

function shp = plotcurve(shp)
    cla;
    p = backbone(shp.system.Backbone);
    plot(p(:,1), p(:,3),'k-','LineW',3);
    axis equal;
    axis([-100, 100, -100, 100]*1.2);
    drawnow;
end
```

**Comment by BJCaasenbrood** [comment]
@Beta-y There are still some things that I need to polish out, and I also need to make sure it is compatible with the other examples of the toolkit. I noticed there is still a bug for when `base_dof(4:6) != 0` (i.e., the beam can move freely in 3D space) -- the free-floating origin of the beam does not to seem evolve naturally in $\mathbb{R}^3$, but considering only rotation (i.e., its evolution on the $SO(3)$ manifold) seems to work fine! I also noticed that twist `base_dof(1) != 0` destabilizes the system quickly -- I will fix this at some point.

Nevertheless, here is an pre-release version. Make sure to compile the new mex files `sorotokimodel/mex/freebodyforwardkin/buildMexFFKFG.m` and `sorotokimodel/mex/freebodylagrangian/buildMexCFLGF.m`

The model now has joint coordinates $[q_1, ..., q_M, \theta_1]$ when `dof_beam = [0,M,0,0,0,0]` and `dof_base = [0,1,0,0,0,0]`. The DOFs of the base are always added at the end.

[prereleaseFreeShapes.zip](https://github.com/BJCaasenbrood/SorotokiCode/files/15146249/prereleaseFreeShapes.zip)

Please let me know if it works for you.

**Comment by BJCaasenbrood** [comment]
@Beta-y Below you'll find the code and result for making an unactuated soft pendulum, you may change `@Control` to a desired stabilizing controller. Let me know if the code below works for you :wink: 

![soft_pendulum_active](https://github.com/BJCaasenbrood/SorotokiCode/assets/36262141/200c4d60-bab1-4267-b6ac-9d8ee193ddc1)

```matlab
M = 12;
mat = NeoHookean(0.1,0.4);
mat.params.Zeta = 0.2;

shp = Shapes(chebyspace(50,M), [0,M,0,0,0,0], [0,1,0,0,0,0]);
shp = shp.setRadius([8,8,0.85]);
shp = shp.setMaterial(mat);
shp = shp.setBase('-z');
shp = shp.rebuild();

shp.solver.TimeStep = 1/160;
shp.solver.TimeHorizon = 5;

shp.options.Display = @plotcurve;
shp = shp.addGravity([0;0;-9800]);

shp = shp.addControl(@Control);

ii = 1;
while shp.solver.Time < shp.solver.TimeHorizon
    shp = shp.update();
    plotcurve(shp);
    if ii == 1
        gif('soft_pendulum_active.gif','nodither',...
            'DelayTime', 1/60,...
            'frame',gca);
    else
        gif;
    end
    ii =  ii + 1;
end

function tau = Control(shp)
    t = shp.solver.Time;
    tau = zeros(shp.NJoint,1);
    tau(end) = .75 * sin(6*pi*t);
end

function shp = plotcurve(shp)
    cla;
    p = backbone(shp.system.Backbone);
    plot(p(:,1), p(:,3),'k-','LineW',3);
    axis equal;
    axis([-100, 100, -100, 100]*1.2);
    drawnow;
end
```

**Reply by Beta-y** [reply]
Hi, Dr.! I have tested the attached zip file. It seems like some properties of `shapesoptions` and `beamoptions` classes are lost, such as `NBase;  BaseTable;  NBaseDof;` and `BaseDofMap`, which are defined in dir named `>sorotokimodel` and not provided in zip file yet. I have added them by myself. And now it works perfectly. I will try to add some closed-loop controller and reply as soon as possible. Thank you a lot for your kind help! ❤

**Reply by BJCaasenbrood** [reply]
@Beta-y here are the missing files: [beam_shapeoptions.zip](https://github.com/BJCaasenbrood/SorotokiCode/files/15146970/beam_shapeoptions.zip)

**Edit**: Nevermind, saw u added them yourself.
